src/engine/SortedUnionImpl.h

```diff
@@ -0,0 +1,241 @@
+//   Copyright 2025, University of Freiburg,
+//   Chair of Algorithms and Data Structures.
+//   Author: Robin Textor-Falconi <textorr@informatik.uni-freiburg.de>
+
+#pragma once
+
+#include <optional>
+
+#include "engine/LocalVocab.h"
+#include "engine/Result.h"
+#include "engine/Union.h"
+#include "engine/idTable/IdTable.h"
+
+// Helper struct that has the same layout as Result::IdTableVocabPair but
+// doesn't own the data.
+struct Wrapper {
+  const IdTable& idTable_;
+  const LocalVocab& localVocab_;
+};
+
+// Move the data from a real `Result::IdTableVocabPair`.
+inline Result::IdTableVocabPair moveOrCopy(Result::IdTableVocabPair& element) {
+  return std::move(element);
+}
+
+// Copy the data from a `Wrapper`.
+inline Result::IdTableVocabPair moveOrCopy(const Wrapper& element) {
+  return {element.idTable_.clone(), element.localVocab_.clone()};
+}
+
+// Range that performs a zipper merge of two sorted ranges.
+template <size_t SPAN_SIZE, typename Range1, typename Range2, typename Func>
+struct SortedUnionImpl
+    : ad_utility::InputRangeFromGet<Result::IdTableVocabPair> {
+  // Iterator and range storage.
+  std::shared_ptr<const Result> result1_;
+  std::shared_ptr<const Result> result2_;
+  Range1 range1_;
+  std::optional<typename Range1::iterator> it1_ = std::nullopt;
+  size_t index1_ = 0;
+  Range2 range2_;
+  std::optional<typename Range2::iterator> it2_ = std::nullopt;
+  size_t index2_ = 0;
```

Comment:
Those are two objects of a (templated) struct that each consists of
* Result (only for keepalive)
* the actual rangeOfTables (for the outer iteration)
* the `tableIterator_` (for the outer iteration).
* the `rowIndex_`  for the inner iteration.

Then here you only have
`IterationData<Range1> iterators1 { result1, range1};`
Same for Range2.

Reply:
Maybe even pass this in as a single element in the constructor.